docs/rfc/0001-exporter-interface.md

```diff
@@ -0,0 +1,194 @@
+maintained either publicly or privately  by them. This plugin specification will also give us a runtime framework and
+configuration necessary for creating other plugin systems, such as input plugins and intermediate plugins.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+### Indexer-Native Exporter Plugins  
+Exporter plugins that are native to the Indexer (maintained within the Indexer repository) will necessarily behave
+differently than those which are maintained and built separately and loaded into the pipeline at runtime. The following
+section will outline the properties of so called "Indexer-native" plugins, which will then be extended in the subsequent
+section to incorporate non-native plugins.
+
+***Plugin Interface***  
+Exporter plugins that are native to the Indexer (maintained within the Indexer repository) will each have the
+following components:
+* An implementation of the exporter interface.
+```
+// Exporter defines the methods invoked during the plugin lifecycle of a data exporter.
+type Exporter interface {
+  // Connect will be called during initialization, before block data starts going through the pipeline.
+  // Typically used for things like initializating network connections.
+  // The ExporterConfig passed to Connect will contain the Unmarhsalled config file specific to this plugin.
+  // Should return an error if it fails--this will result in the Indexer process terminating.
+  Connect(ExporterConfig cfg) error
+  
+  // Shutdown will be called during termination of the Indexer process.
+  // There is no guarantee that plugin lifecycle hooks will be invoked in any specific order in relation to one another.
+  // Returns an error if it fails which will be surfaced in the logs, but the process is already terminating.
+  Shutdown() error
+  
+  // Recv is called for each block to be processed by the exporter.
+  // Should return an error on failure--retries are configurable.
+  Recv(DataFormat blockData) error
+  
+  // DataFormat returns the type of the data that the plugin expects to receive.
+  // Must be one of the impelemented formats in the Indexer repository.
+  DataFormat() DataFormat
+  
+  // Round returns the next round to be processed. Atomically updated when Recv successfully completes.
+  Round() uint64
+}
+```
+
+***Plugin Config***
+* A config file that defines all parameters that can be supplied to the plugin, and which provides the default values
+that will be used for each parameter. A toml file stored inside the Indexer data directory stores all of the config data
+for a given plugin. The Indexer will use the type specified in the Indexer config to look for a plugin config which
+satisfies that class, and then load that as the selected plugin. Supplying multiple plugin configs for the selected type
+of exporter will result in undefined behavior--a random config will be chosen.
```

Comment:
* TOML: I don't have a strong opinion about these formats, but we already use YML, so I'd opt to stick with it.
* Undefined behavior: let's generate an error, or support multiple instances of the same plugin.